src/general_functions.jl

```diff
@@ -311,6 +311,37 @@ function expect(
     return ishermitian(O) ? real(tr(O * ρ)) : tr(O * ρ)
 end
 
+@doc raw"""
+    fidelity(ρ::QuantumObject, σ::QuantumObject)
+
+Calculate the fidelity of two [`QuantumObject`](@ref):
+``F(\rho, \sigma) = \textrm{Tr} \sqrt{\sqrt{\rho} \sigma \sqrt{\rho}}``
+
+Here, the definition is from Nielsen & Chuang, "Quantum Computation and Quantum Information". It is the square root of the fidelity defined in R. Jozsa, Journal of Modern Optics, 41:12, 2315 (1994).
+
+Note that `ρ` and `σ` must be either [`Ket`](@ref) or [`Operator`](@ref).
+"""
+function fidelity(
+    ρ::QuantumObject{<:AbstractArray{T1},OperatorQuantumObject},
+    σ::QuantumObject{<:AbstractArray{T2},OperatorQuantumObject},
+) where {T1,T2}
+    sqrt_ρ = sqrt(ρ)
+    eigval = abs.(eigvals(sqrt_ρ * σ * sqrt_ρ))
+    return sum(sqrt, eigval)
+end
+fidelity(
+    ρ::QuantumObject{<:AbstractArray{T1},OperatorQuantumObject},
+    ψ::QuantumObject{<:AbstractArray{T2},KetQuantumObject},
+) where {T1,T2} = sqrt(abs(expect(ρ, ψ)))
+fidelity(
+    ψ::QuantumObject{<:AbstractArray{T1},KetQuantumObject},
+    σ::QuantumObject{<:AbstractArray{T2},OperatorQuantumObject},
+) where {T1,T2} = fidelity(σ, ψ)
+fidelity(
+    ψ::QuantumObject{<:AbstractArray{T1},KetQuantumObject},
+    ϕ::QuantumObject{<:AbstractArray{T2},KetQuantumObject},
+) where {T1,T2} = abs(ψ' * ϕ)
```

Comment:
Should we perform `abs(dot(a, b))` instead?

Reply:
sure~